src/probability/kernel/composition.lean

```diff
@@ -0,0 +1,567 @@
+  prod_fun κ η a s = ∑' n, prod_fun κ (seq η n) a s :=
+begin
+  simp_rw [prod_fun, (measure_sum_seq η _).symm],
+  have : ∫⁻ b, measure.sum (λ n, seq η n (a, b)) {c : γ | (b, c) ∈ s} ∂(κ a)
+    = ∫⁻ b, ∑' n, seq η n (a, b) {c : γ | (b, c) ∈ s} ∂(κ a),
+  { congr',
+    ext1 b,
+    rw measure.sum_apply,
+    exact measurable_prod_mk_left hs, },
+  rw [this, lintegral_tsum (λ n : ℕ, _)],
+  exact ((measurable_prod_mk_mem (seq η n) ((measurable_fst.snd.prod_mk measurable_snd) hs)).comp
+    measurable_prod_mk_left).ae_measurable,
+end
+
+lemma prod_fun_tsum_left (κ : kernel α β) (η : kernel (α × β) γ) [is_s_finite_kernel κ]
+  (a : α) (s : set (β × γ)) :
+  prod_fun κ η a s = ∑' n, prod_fun (seq κ n) η a s :=
+by simp_rw [prod_fun, (measure_sum_seq κ _).symm, lintegral_sum_measure]
+
+lemma prod_fun_eq_tsum (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] (a : α) (hs : measurable_set s) :
+  prod_fun κ η a s = ∑' n m, prod_fun (seq κ n) (seq η m) a s :=
+by simp_rw [prod_fun_tsum_left κ η a s, prod_fun_tsum_right _ η a hs]
+
+/-- Auxiliary lemma for `measurable_prod_fun`. -/
+lemma measurable_prod_fun_of_finite (κ : kernel α β) [is_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_finite_kernel η] (hs : measurable_set s) :
+  measurable (λ a, prod_fun κ η a s) :=
+begin
+  simp only [prod_fun],
+  have h_meas : measurable (function.uncurry (λ a b, η (a, b) {c : γ | (b, c) ∈ s})),
+  { have : function.uncurry (λ a b, η (a, b) {c : γ | (b, c) ∈ s})
+      = λ p, η p {c : γ | (p.2, c) ∈ s},
+    { ext1 p,
+      have hp_eq_mk : p = (p.fst, p.snd) := prod.mk.eta.symm,
+      rw [hp_eq_mk, function.uncurry_apply_pair], },
+    rw this,
+    exact measurable_prod_mk_mem η (measurable_fst.snd.prod_mk measurable_snd hs), },
+  exact measurable_lintegral κ h_meas,
+end
+
+lemma measurable_prod_fun (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] (hs : measurable_set s) :
+  measurable (λ a, prod_fun κ η a s) :=
+begin
+  simp_rw prod_fun_tsum_right κ η _ hs,
+  refine measurable.ennreal_tsum (λ n, _),
+  simp only [prod_fun],
+  have h_meas : measurable (function.uncurry (λ a b, seq η n (a, b) {c : γ | (b, c) ∈ s})),
+  { have : function.uncurry (λ a b, seq η n (a, b) {c : γ | (b, c) ∈ s})
+      = λ p, seq η n p {c : γ | (p.2, c) ∈ s},
+    { ext1 p,
+      have hp_eq_mk : p = (p.fst, p.snd) := prod.mk.eta.symm,
+      rw [hp_eq_mk, function.uncurry_apply_pair], },
+    rw this,
+    exact measurable_prod_mk_mem (seq η n) (measurable_fst.snd.prod_mk measurable_snd hs), },
+  exact measurable_lintegral κ h_meas,
+end
+
+/-- Product of kernels. It verifies
+`∫⁻ bc, f bc.1 bc.2 ∂(prod κ η a) = ∫⁻ b, ∫⁻ c, f b c ∂(η (a, b)) ∂(κ a)` (see `lintegral_prod`). -/
+noncomputable
+def prod (κ : kernel α β) [is_s_finite_kernel κ] (η : kernel (α × β) γ) [is_s_finite_kernel η] :
+  kernel α (β × γ) :=
+{ val := λ a, measure.of_measurable (λ s hs, prod_fun κ η a s) (prod_fun_empty κ η a)
```

Comment:
The name `prod` puzzles me a bit; I would expect it to correspond to [`measure.prod`](https://leanprover-community.github.io/mathlib_docs/measure_theory/constructions/prod.html#measure_theory.measure.prod), but obviously it doesn't.

Is that meaning of `prod` not one that is useful on kernels? If it is, then it might make sense to leave the name available and use a different name for this.

Reply:
You're right. I need to find another name. The only name I found for this construction is "composition", but it clashes with the other composition.

Reply:
I just remembered why I called this operation product: it is the name used on the Wikipedia page for transition kernels (https://en.wikipedia.org/wiki/Transition_kernel).

The direct generalization of `measure.prod` to kernels we get by seeing measures as constant kernels from some common space `α`  would be a definition `kernel α β → kernel α γ → kernel α (β × γ)`, which is the particular case of the product I defined here in which the second argument does not depend on `β`.

Reply:
It has both flavors of a composition and of a product, so why not call it a composition product? And denote it with `comp_prod` in lemma names? Like @eric-wieser, I'd rather keep `prod` for the genuine product of kernels. Anyway, since you use a notation most of the time, it shouldn't make a big difference.

Reply:
I changed the name to `comp_prod` and also added a new `prod` for the product.
The PR is now quite big as a result. I could split it again if you prefer.

Reply:
Hmm, I expected `prod` to be defined in term of `measure.prod`; I guess it can't be split so easily given it's not.